src/Compilers/Core/Portable/Collections/ImmutableArrayExtensions.cs

```diff
@@ -497,6 +499,22 @@ public static int Count<T>(this ImmutableArray<T> items, Func<T, bool> predicate
             return count;
         }
 
+        private static class UnderlyingArrayHelper<T>
+        {
+            public static readonly Func<object, object> GetValue =
+                typeof(ImmutableArray<T>).GetField("array",
+                    BindingFlags.Instance | BindingFlags.NonPublic).GetValue;
+        }
+
+        internal static T[] DangerousGetUnderlyingArray<T>(this ImmutableArray<T> array)
+            => (T[])UnderlyingArrayHelper<T>.GetValue(array);
```

Comment:
>(T[])UnderlyingArrayHelper<T>.GetValue(array); [](start = 15, length = 46)

Do we assume that the price we pay for going through reflection will be offset by something else?

Reply:
My _guess_ is we add support for the `ReadOnlySpan<T>` and `ReadOnlyMemory<T>` conversions to the underlying library, and the use of reflection is only an intermediate state.

Reply:
That's correct. I've added a comment that points to the bug where we can get rid of this.